packages/dds/tree/src/util/utils.ts

```diff
@@ -54,6 +54,18 @@ export function fail(message: string): never {
 	throw new Error(message);
 }
 
+/**
+ * Use this function to assert that an array index is not out-of-bounds.
+ * @example
+ * ```ts
+ * // We know that `numberArray` has four elements in it, so this is safe.
+ * const n: number = numberArray[3] ?? oob();
+ * ```
+ */
+export function oob(): never {
+	return fail("Array index is out of bounds");
+}
```

Comment:
What about unintentionally sparse arrays? Should the failure message always be about "out of bounds"? I feel like, if this is going to be hit, then it is more likely to be a rogue sparse array (assuming we've actually done the bounds check we say either through inspection or a loop).

Reply:
It's a good thought. This should only be used for unsparse (dense?) arrays. So, any array that is typed as `T[]` where `T` does not include `undefined`. It would not be appropriate to use `?? oob()` for an array of `(number | undefined)[]` for example. The doc comment here could call this out as a remark for extra clarity.

For _unintentionally_ sparse arrays, where an array is typed as e.g. `number[]` but it has some unexpected `undefined`s in it, then using `oob` would be problematic (and so would `assert`ing, and so would using `!`). I think it's a lot harder to end up with an accidentally sparse array than it is to, say, read past the end of an array, so my intuition is that as a rule we can assume a `T[]` contains only `T`s and not `T | undefined`s.

By the way, I was mentioning to Rishhi and he said you had a similar thought - I'd us to have a custom lint rule where we allow `!` to be used for array index access, but not for anything else. If we ever do that, we can go back and replace all `oob`s with `!`.

Reply:
And thinking more about it, I agree with you that the actual error message here should be more generic, like "Expected element to be defined in array" because you're right - we likely won't hit these often, but if we _do_, it very well might be the sparse case since presumably we validated against the out of bounds case.

Reply:
There is a functional difference with `?? oob()` and `!`. The former catches point of failure whereas the latter lets the failure happen at an unspecified later time (in some patterns). `?? oob()` also is more code and impacts bundle size.

It feels like since we have been working with the effective `!` world, then we are okay if we instead switched to lint rule now. Once we start using `?? oob()` it is not clear we always just go back -- unless we are very prescriptive about its use. Kind of like we'd want
```typescript
function safelyIndex<T>(array: Omit<T, undefined | null>[], index: number): T {
   return array[index] ?? fail("Unexpected missing array element or out of bounds");
}
```